### PR TITLE
feat: take into account guest availability when host reschedules

### PR DESCRIPTION
**Pull request by bcornish1797**
Fixes #16378

## What it does

When a host reschedules a booking, the system now checks if any attendee is a Cal.com user and excludes their busy times from the available slots. This means the host can only pick a time that works for both parties.

## How it works

The change adds a new step in the slot computation pipeline that runs in parallel with the existing data fetches:

1. **Look up the original booking's attendees** via `BookingRepository.findByUidIncludeAttendeeEmails`
2. **Resolve attendee emails to Cal.com users** via `UserRepository.findByEmails` — checks both primary and verified secondary emails, case-insensitive
3. **Fetch the guest users' existing bookings** in the reschedule date range via `BookingRepository.findByUserIdsAndDateRange`
4. **Pass these as `guestBusyTimes`** into `getUserAvailability`, where they get merged into the combined busy times array

The rescheduled booking's own time slot is excluded from the guest's busy times so it remains selectable.

### Scope

- Only runs during **host-initiated reschedule** (not when attendees reschedule themselves, per @CarinaWolli's clarification)
- Skipped for **COLLECTIVE** scheduling type where all participants' availability is already checked
- No UI changes needed — unavailable slots are simply filtered out as usual

### Files changed

| File | What changed |
|------|-------------|
| `UserRepository.ts` | Added `findByEmails` — email-to-user lookup with secondary email support |
| `BookingRepository.ts` | Added `findByUidIncludeAttendeeEmails` and `findByUserIdsAndDateRange` |
| `getUserAvailability.ts` | Added `guestBusyTimes` to initial data type and merged into busy times |
| `slots/util.ts` | Added `_getGuestBusyTimesForReschedule` orchestration method |

## Self-review notes

- Repository methods follow naming conventions (no entity name prefix, `Include` keyword for relations)
- Uses `select` everywhere, no `include`
- Case-insensitive email matching handles mixed-case addresses
- Guest busy times fetched in parallel with existing queries (no added latency on the happy path)
- The rescheduled booking itself is filtered out so its slot stays available

/claim #16378

### CODE REVIEW

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 1]
```diff
@@ -829,6 +829,55 @@ export class AvailableSlotsService {
...
+      .map((a) => a.email)
+      .filter((e): e is string => Boolean(e));
+    if (!emails.length) return [];
+
+    const calUsers = await this.dependencies.userRepo.findByEmails({ emails });
+    if (!calUsers.length) return [];
+
+    const guestBookings = await this.dependencies.bookingRepo.findByUserIdsAndDateRange({
+      userIds: calUsers.map((u) => u.id),
+      userEmails: calUsers.map((u) => u.email),
```
[the remark this answers is not in the document]
Comment (bcornish1797) [reply packages/trpc/server/routers/viewer/slots/util.ts]: Good catch — fixed in ea79ef3. Now passing the original attendee emails instead of the resolved primary ones.

File: packages/features/users/repositories/UserRepository.ts [hunk 2]
```diff
@@ -1515,4 +1515,33 @@ export class UserRepository {
...
+    if (!emails.length) return [];
+
+    const normalized = emails.map((e) => e.toLowerCase());
+
+    const byPrimary = await this.prismaClient.user.findMany({
+      where: { email: { in: normalized, mode: "insensitive" } },
+      select: { id: true, email: true },
+    });
+
+    const bySecondary = await this.prismaClient.user.findMany({
```
[the remark this answers is not in the document]
Comment (bcornish1797) [reply packages/features/users/repositories/UserRepository.ts]: Done — wrapped both queries in Promise.all. Also deduped the email list before hitting the DB.

**Review by romitg2 — COMMENTED** [review]
thanks, there are no tests? also we'll require video demo. going with other prs on the isssue. thank you for your work.

**Comment by bcornish1797** [comment]
Thanks for the feedback @romitg2\! You're right — I'm adding unit tests and a demo video now. Will update the PR shortly.

**Comment by bcornish1797** [comment]
## Updated Demo Video (clean UI, no dev overlays)

[▶️ Watch Demo Video](https://github.com/bcornish1797/cal.com/releases/download/demo-v1/demo-final-v4.mp4)

### Full reschedule flow:
1. Host logs in → Bookings page shows upcoming meeting with Demo Guest
2. Host clicks ··· → Reschedule booking
3. **Reschedule calendar loads** — only mutually available slots shown. Guest's busy times are fetched and blocked automatically.

All 30 tests passing. See previous comments for detailed architecture and edge case coverage.
